server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1074,35 +1083,29 @@ public void onFailure(String source, Exception e) {
                 if (runningSnapshot == null) {
-                    threadPool.generic().execute(ActionRunnable.wrap(listener, l ->
-                        repositoriesService.repository(repositoryName).getRepositoryData(
-                            ActionListener.wrap(repositoryData -> {
-                                Optional<SnapshotId> matchedEntry = repositoryData.getSnapshotIds()
-                                    .stream()
-                                    .filter(s -> s.getName().equals(snapshotName))
-                                    .findFirst();
-                                // If we can't find the snapshot by the given name in the repository at all or if the snapshot we find in
-                                // the repository is not the one we expected to find when waiting for a finishing snapshot we fail.
-                                if (matchedEntry.isPresent()) {
-                                    deleteCompletedSnapshot(
-                                        new Snapshot(repositoryName, matchedEntry.get()), repositoryData.getGenId(), Priority.NORMAL, l);
-                                } else {
-                                    l.onFailure(new SnapshotMissingException(repositoryName, snapshotName));
-                                }
-                            }, l::onFailure))));
+                    resolveAndDeleteSnapshots(snapshotNames, listener, repositoryName);
                     return;
                 }
                 logger.trace("adding snapshot completion listener to wait for deleted snapshot to finish");
                 addListener(runningSnapshot, ActionListener.wrap(
                     result -> {
                         logger.debug("deleted snapshot completed - deleting files");
-                        deleteCompletedSnapshot(
-                            new Snapshot(repositoryName, result.v2().snapshotId()), result.v1().getGenId(), Priority.IMMEDIATE, listener);
+                        final RepositoryData newRepoData = result.v1();
+                        final List<SnapshotId> foundSnapshots = matchingSnapshotIds(newRepoData, snapshotNames, repositoryName);
+                        assert foundSnapshots.contains(runningSnapshot.getSnapshotId()) : "Expected new RepositoryData to contain ["
+                                + runningSnapshot + "] but it only contained snapshots " + newRepoData.getSnapshotIds();
+                        deleteCompletedSnapshots(foundSnapshots, repositoryName, newRepoData.getGenId(), Priority.IMMEDIATE, listener);
                     },
                     e -> {
                         if (abortedDuringInit) {
                             logger.info("Successfully aborted snapshot [{}]", runningSnapshot);
-                            listener.onResponse(null);
+                            final Collection<String> remainingPatterns = new ArrayList<>(snapshotNames);
```

Comment:
There is a bit of weirdness here in that we abort the matching snapshot before we check for existing snapshots in the repo. We will have to discuss what exact mechanics we want for the multi-delete when it partially matches.
The easiest to implement would be to ignore missing snapshots and maybe just either throw a `404` if nothing is matched for a delete or go with it if anything matches.
The current approach here might be ok but not great though, in that the abort will happen but the subsequent delete might still fail with a `404` because not all the snapshots are found.

Another approach may be to only allow aborts by exact name in a single snapshot delete request for now and deal with this in the next step. Eventually this issue can be resolved cleanly by adding another step to the snapshot delete operation, adding a placeholder `Entry` for the delete before actually running it to prevent concurrent snapshots from modifying the repository between reads (exactly like we do with the snapshot `INIT` state before checking that the given snapshot name is available in the repo) but this PR is already pretty long and I'm not sure we want to do that here already as it request a number of changes to master fail-over handling and such.